internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,203 @@
+package spectest
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	"github.com/reddit/baseplate.go/errorsbp"
+)
+
+const (
+	client = "client"
+	server = "server"
+)
+
+var (
+	errPrefix         = errors.New("the prefix is not at the beginning of the metric name")
+	errClientServer   = fmt.Errorf("missing %q or %q as the second part of the metric name", client, server)
+	errLength         = errors.New("metric name should have a minimum of three parts")
+	errNotFound       = errors.New("metric not found")
+	errLabelNotFound  = errors.New("label not found")
+	errPrometheusLint = errors.New("prometheus GatherAndLint")
+)
+
+// ValidateSpec validates that the Prometheus metrics being exposed from baseplate.go
+// conform to the baseplate spec.
+func ValidateSpec(t *testing.T, metricPrefix string, wantMetricCount int) {
+	t.Helper()
+
+	var batch errorsbp.Batch
+	batch.AddPrefix("validate spec", validateSpec(metricPrefix))
+
+	if len(batch.GetErrors()) > 0 {
+		t.Errorf(batch.Compile().Error())
+	}
+}
+
+func validateSpec(prefix string) errorsbp.Batch {
+	var batch errorsbp.Batch
+
+	metricFam, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		return batch
```

Comment:
do you mean `return err`, or `batch.Add(err); return batch.Compile()` here?

Reply:
good catch. I will `batch.Add(err); return batch`

Reply:
this is nitpick/optional, but I would prefer to make this function to return `error` instead of `errorsbp.Batch` (so you just add `.Compile()` to every return inside this function). `batch.AddPrefix` on line 38 can correctly convert an `error` back to a batch when needed.

Reply:
originally I returned error, but when talking with Kyle we changed it to return errorsbp.Batch because it make testing tricky because then I can't test the error itself, but rather have to do string matching which i would like to avoid if possible. ref: https://github.com/reddit/baseplate.go/pull/445/files#diff-8b8f4f423ed61950357bb8113cc700075e8a2a31b64b6a6b6bb878b774fdf5faR288

Reply:
OK in that case this is fine.

Reply:
but compiled batch errors still implemented `errors.Is`/`errors.As` correctly, so that should not really affect those tests. (e.g. if you change the signature of this function to return `error` instead, the tests should still pass).

Reply:
I originally tried `errors.As` but it didn't work, but let me try again since a lot has changed since i was attempting that.